### PR TITLE
 BUG: datetime.date objects don't get parsed for non-ISO formats

### LINKED ISSUE #50108: BUG: datetime.date objects don't get parsed for non-ISO formats
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
to_datetime([dt.date(2000, 1, 2)], format='%Y-%m-%d %H:%M:%S%z')  # works!
to_datetime([dt.date(2000, 1, 2)], format='%Y-%d-%m %H:%M:%S%z')  # fails!
```

### Issue Description

```
Traceback (most recent call last):
  File "t.py", line 45, in <module>
    to_datetime([dt.date(2000, 1, 2)], format='%Y-%d-%m %H:%M:%S%z')
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 1112, in to_datetime
    result = convert_listlike(argc, format)
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 438, in _convert_listlike_datetimes
    res = _to_datetime_with_format(
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 544, in _to_datetime_with_format
    res = _array_strptime_with_fallback(
  File "/home/marcogorelli/pandas-dev/pandas/core/tools/datetimes.py", line 478, in _array_strptime_with_fallback
    result, timezones = array_strptime(
  File "pandas/_libs/tslibs/strptime.pyx", line 189, in pandas._libs.tslibs.strptime.array_strptime
    raise ValueError(f"time data '{val}' does not match "
ValueError: time data '2000-01-02' does not match format '%Y-%d-%m %H:%M:%S%z' (match)
```

### Expected Behavior

```
DatetimeIndex(['2000-01-02'], dtype='datetime64[ns]', freq=None)
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : e93ee07729afe0bc7661655755df6adad657c23b
python           : 3.8.15.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.0.dev0+863.ge93ee07729
numpy            : 1.23.5
pytz             : 2022.6
dateutil         : 2.8.2
setuptools       : 65.5.1
pip              : 22.3.1
Cython           : 0.29.32
pytest           : 7.2.0
hypothesis       : 6.59.0
sphinx           : 4.5.0
blosc            : None
feather          : None
xlsxwriter       : 3.0.3
lxml.etree       : 4.9.1
html5lib         : 1.1
pymysql          : 1.0.2
psycopg2         : 2.9.3
jinja2           : 3.0.3
IPython          : 8.7.0
pandas_datareader: 0.10.0
bs4              : 4.11.1
bottleneck       : 1.3.5
brotli           : 
fastparquet      : 2022.11.0
fsspec           : 2021.11.0
gcsfs            : 2021.11.0
matplotlib       : 3.6.2
numba            : 0.56.4
numexpr          : 2.8.3
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : 0.17.9
pyarrow          : 9.0.0
pyreadstat       : 1.2.0
pyxlsb           : 1.0.10
s3fs             : 2021.11.0
scipy            : 1.9.3
snappy           : 
sqlalchemy       : 1.4.44
tables           : 3.7.0
tabulate         : 0.9.0
xarray           : 2022.11.0
xlrd             : 2.0.1
zstandard        : 0.19.0
tzdata           : None
qtpy             : None
pyqt5            : None
None

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #50108 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/tslibs/strptime.pyx [hunk 1]
```diff
@@ -173,6 +175,10 @@ def array_strptime(
                 check_dts_bounds(&dts)
             result_timezone[i] = val.tzinfo
             continue
+        elif PyDate_Check(val):
+            iresult[i] = pydate_to_dt64(val, &dts)
+            check_dts_bounds(&dts)
+            continue
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/strptime.pyx:181]: same as

https://github.com/pandas-dev/pandas/blob/7c0278e533b8cf19a8b6bd367bcb26a06d08ab9e/pandas/_libs/tslib.pyx#L544-L547

File: doc/source/whatsnew/v2.0.0.rst [hunk 2]
```diff
@@ -698,7 +698,7 @@ Datetimelike
 - Bug in subtracting a ``datetime`` scalar from :class:`DatetimeIndex` failing to retain the original ``freq`` attribute (:issue:`48818`)
 - Bug in ``pandas.tseries.holiday.Holiday`` where a half-open date interval causes inconsistent return types from :meth:`USFederalHolidayCalendar.holidays` (:issue:`49075`)
 - Bug in rendering :class:`DatetimeIndex` and :class:`Series` and :class:`DataFrame` with timezone-aware dtypes with ``dateutil`` or ``zoneinfo`` timezones near daylight-savings transitions (:issue:`49684`)
-- Bug in :func:`to_datetime` was raising ``ValueError`` when parsing :class:`Timestamp`, ``datetime``, or ``np.datetime64`` objects with non-ISO8601 ``format`` (:issue:`49298`, :issue:`50036`)
+- Bug in :func:`to_datetime` was raising ``ValueError`` when parsing :class:`Timestamp`, ``datetime.datetime``, ``datetime.date``, or ``np.datetime64`` objects with non-ISO8601 ``format`` (:issue:`49298`, :issue:`50036`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.0.0.rst:701]: Nit: This makes it sounds like `format` is being applied to these non-string objects (which IIUC is not the case). Could the phrasing be changed to not suggest this?

Comment (MarcoGorelli) [reply doc/source/whatsnew/v2.0.0.rst]: good point - is the new wording clearer?

Comment (mroeschke) [reply doc/source/whatsnew/v2.0.0.rst]: Thanks! That's better

**Review by mroeschke — APPROVED** [review]

File: pandas/_libs/tslibs/strptime.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx:181]: do we need to worry about found_naive?  (i guess we dont currently in the tslib code?)

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/strptime.pyx:181]: Not sure we need to, `datetime.date`s can't have a timezone, so perhaps it's fine to just parse?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/strptime.pyx:181]: OK to punt on this for now

**Pull request merged**